### ISSUE TITLE
Home Assistant pages fail  to load due to CORS policy / security errors

### ISSUE THREAD

**Issue by mprobst** [home-assistant/frontend#19911]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.
- [ ] I have tried reproducing the issue in [safe mode](https://www.home-assistant.io/blog/2023/11/01/release-202311/#restarting-into-safe-mode) to rule out problems with unsupported custom resources.

### Describe the issue you are experiencing

Various sub-pages in Home Assistant fail to load on my installation. It appears that some CORS header isn't set correctly in the server response when dynamically loading additional JavaScript snippets.

![Screenshot 2024-02-28 200921](https://github.com/home-assistant/frontend/assets/7208/d4d17a99-5059-45a1-894b-6b272028171a)

Note the log entries:

```
Access to script at 'http://lokalhorst:8123/frontend_latest/44356.ka0DMKv3Wu4.js' from origin 'http://lokalhorst:8123' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `private`.
44356.ka0DMKv3Wu4.js:1 
Failed to load resource: net::ERR_FAILED
```

This happens in latest Chrome (121.0.6167.185), but not (yet?) in latest Firefox (123.0).

### Describe the behavior you expected

Should load the pages on navigation.

### Steps to reproduce the issue

0. Clear all browser caches.
1. Navigate to my Home Assistant (http://lokalhorst:8123, i.e. non-SSL, using a hostname. Not sure if that's required)
2. Navigate to a sub-page, e.g. media or energy
3. Observe the failure

Strangely enough this appears to only happen on first attempt (and then persistently), but when I close the tab and try again, things work. Maybe there's some caching effect?

### What version of Home Assistant Core has the issue?

core-2024.2.4

### In which browser are you experiencing the issue with?

Google Chrome 121.0.6167.185

### Which operating system are you using to run this browser?

Windows 11

**Comment by WarMasterNL** [comment]
I have the same issue with increasing frequency. Weirdly not so many people with the same issue it seems.

**Comment by longzheng** [comment]
I also run into this infrequently. 

Today when I saw the error, I took a quick look at Chrome DevTools and I saw that it was resolving my Home Assistant domain using a IPv6 address. But when I opened the Home Assistant domain in an incognito window which loaded correctly, it was resolving using a different IPv6 link-local `[fe80::` address.

I didn't have a lot of time to investigate before my normal Chrome window started working again, but it makes me wonder if it's something related to IPv6 addresses.

**Comment by thromera** [comment]
I'm having that issue when accessing Homeassistant locally with `http://homeassistant.local:8123/`. 

```
Access to fetch at 'http://homeassistant.local:8123/auth/login_flow' from origin 'http://homeassistant.local:8123' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `private`.
```

Weirdly, after a few hard refreshes, it sometimes disappear. 
Sometimes, a page works, then you click on another page, and the error is back on that page until hard refresh works at some point.

**Comment by rkboni** [comment]
I just had this happen to me (and have seen it happen before but it always self-resolved rather quickly), and I'm wondering if it has something to do with DHCP registration / DNS reachability, as I rebooted my router (also my DNS/DHCP server) to apply a software upgrade shortly before this happened.

My HA is registered as `homeassistant.home` in my local DNS, and 99% of the time that works just fine.  After my reboot of the router, navigating various HA pages worked, but for whatever reason the ESPHome dashboard refused to load when accessed via http://homeassistant.home:8123/5c53de3b_esphome (what the sidebar links to).  Errors in the Chrome console were the same as vein as those quoted at the top of the ticket:

```
 Access to script at 'http://homeassistant.home:8123/...' from origin 'http://homeassistant.home:8123' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `private`.
```

My HA system also has an IPv6 address, but I realized that the DNS name is only associated with the IPv4 address... so it might be that as a result of the DHCP/DNS server reboot the HA "loses" IPv4 connectivity temporarily even if the browser and other end systems still have the name cached?
